src/Faction/Factions.ts

```diff
@@ -2,63 +2,47 @@
+import { Augmentations } from "../Augmentation/Augmentations";
+import { getEnumHelper } from "../utils/EnumHelper";
 
-export let Factions: Record<string, Faction> = {};
-
-export function loadFactions(saveString: string): void {
-  Factions = JSON.parse(saveString, Reviver);
-  // safety check for when we load older save file that don't have newer factions
-  for (const faction of Object.keys(Factions)) {
-    try {
-      Factions[faction].getInfo();
-    } catch (err) {
-      console.error("deleting " + faction);
-      delete Factions[faction];
-    }
+/** The static list of all factions. Initialized once and never modified. */
+export const Factions = createEnumKeyedRecord(FactionName, (name) => new Faction(name));
+// Add the associated augs to every faction
+for (const aug of getRecordValues(Augmentations)) {
+  for (const factionName of aug.factions) {
+    const faction = Factions[factionName];
+    faction.augmentations.push(aug.name);
   }
 }
 
-function AddToFactions(faction: Faction): void {
-  const name: string = faction.name;
-  Factions[name] = faction;
-}
-
-export function factionExists(name: string): boolean {
-  return Object.hasOwn(Factions, name);
-}
-
-export function initFactions(): void {
-  for (const name of Object.keys(FactionInfos)) {
-    resetFaction(new Faction(name));
-  }
-  // All factions are added, this is a good place to add augs back to factions.
-  initCircadianModulator();
-  for (const aug of getRecordValues(Augmentations)) {
-    for (const factionName of aug.factions) {
-      const faction = Factions[factionName];
-      if (!faction) {
-        console.error(`Faction ${factionName} did not exist while adding augs to factions`);
-        continue;
-      }
-      faction.augmentations.push(aug.name);
+export function loadFactions(saveString: string): void {
+  // The only information that should be loaded from player save is
+  const loadedFactions = JSON.parse(saveString, Reviver) as unknown;
+  // This loading method allows invalid data in player save, but just ignores anything invalid
+  if (!loadedFactions) return;
+  if (typeof loadedFactions !== "object") return;
+  for (const [loadedFactionName, loadedFaction] of Object.entries(loadedFactions) as [string, unknown][]) {
+    if (!getEnumHelper("FactionName").isMember(loadedFactionName)) continue;
+    if (!loadedFaction) continue;
+    const faction = Factions[loadedFactionName];
+    if (typeof loadedFaction !== "object") continue;
+    if ("favor" in loadedFaction && typeof loadedFaction.favor === "number" && loadedFaction.favor > 0) {
+      faction.favor = loadedFaction.favor;
     }
+    if (
+      "playerReputation" in loadedFaction &&
+      typeof loadedFaction.playerReputation === "number" &&
+      loadedFaction.playerReputation > 0
+    ) {
+      faction.playerReputation = loadedFaction.playerReputation;
+    }
+    if ("alreadyInvited" in loadedFaction && loadedFaction.alreadyInvited) faction.alreadyInvited = true;
```

Comment:
This entire set of checks would be much less clunky if you did a cast earlier. All the "in" tests are just there to satisfy TS; we're perfectly fine doing a Boolean test against undefined.
I think you want to cast it to a `Record<string, unknown>` instead.

Reply:
It's also clunky finding a good place to do that casting. I can't do it in the for statement because at that point, it could also be `undefined` or `null` (well really just `null` because `undefined` wouldn't be returned by parsing JSON, but ts doesn't know that), and checking a property on something nullish throws an error.

I think I'll make an assertion function that takes in an object and just asserts that it's a better type that's still safe (either Record<string, unknown> or maybe an Unknownify<DesiredType> so that we get autocomplete on keys)

Reply:
> I can't do it in the for statement because at that point, it could also be undefined or null

Huh? By that point, you'd already checked that `loadedFaction` was truthy, so it couldn't be undefined or null. You'd also checked that it was an object. A cast would have been totally safe.

If you want typesafe loading, we'll need a very different structure than what we currently have.

Reply:
The for loop is when loadedFaction is first declared, I have not checked anything about loadedFaction at that point.

I've checked that loadedFactions (the parent object, which is expected to be in the structure of a Record<FactionName, Faction>) is an object, because it makes sense to do that before attempting to iterating through its entries. But those entries have totally unknown values so I cannot assign a type other than unknown inside the for statement.

Reply:
I feel like I'm taking crazy pills.

https://github.com/bitburner-official/bitburner-src/pull/644/commits/0552f73ee05af90e45c96752ebf71e1c0bc8d9c8#diff-796704400cb6ec180a36942fe55e6af741f2d8aa3f8303084324e468028ed4d5

src/Factions/Factions.ts:31
`if (!loadedFaction) continue;`
Line 33:
`if (typeof loadedFaction !== "object") continue;`

The code has changed since then, but in the context of how it was when this thread started, the checks were there already.

Reply:
That is after the for statement.

Reply:
I was just saying that it's a pain to do a manual type reassignment at a time other than when you're declaring the variable (in the for statement declaration, when the correct value is unknown). There are basically these options to reassign the type after-the-fact:

1. Do an assertion function (which I ended up doing) to convert the type of the existing variable
2. Not actually reassign the type, but do `as` statements every time you reference the variable. We would have had a bunch of statements like `if ((loadedFaction as Unknownify<Faction>).isMember) faction.isMember = true`.
3. Assign a new variable with the desired type (e.g. `newLoadedFaction = loadedFaction as Unknownify<Faction>;`) and use the new variable that has the desired type.

Reply:
> If you want typesafe loading, we'll need a very different structure than what we currently have.

I think it's reasonable for things like Factions / Companies / StockMarket where the majority of the data is static and an exact structure can already be generated by the game without needing to load it. In those cases we can just ignore invalid data while loading and we know the result is going to be the correct structure.

But yeah pretty much anything else, not going to be so easy, and I think in those other cases we would want to stop the game from loading. But we could still use a typesafe approach to loading in order to not load invalid data and instead tell the player what is wrong with their save (e.g. a server had a non-string name, or stuff like that), instead of now where we are just loading the invalid data and things may end up with incorrect types and that can cause some random error later on.

Reply:
> There are basically these options to reassign the type after-the-fact:

There is also option #4, which is to reassign the type before-the-fact, and then check after. It's kinda dirty, but it *is* an option. I was actually assuming option 3 in this particular case, but that's because I was talking about a minor change. Using a type assertion function makes more overall sense, especially when it can be reused.

> But yeah pretty much anything else, not going to be so easy, and I think in those other cases we would want to stop the game from loading when we find an error.

This is basically all my comment was about. Right now we're blithely loading the JSON and calling it good. To do it right requires either a lot of copying and checking, or validating it as we deserialize. The former is probably more robust to our format changes. Either way, it's a lot of extra code. Probably worth doing to prevent weird errors (especially given that there's an achievement for explicitly modifying the save file), but still a bunch of work.